File: days/day_03/solutions/solution.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

// Find the largest N-digit number by selecting N digits in order from bank
func findLargestSubsequence(bank string, n int) int64 {
	bankLen := len(bank)
	if bankLen < n {
		return 0
	}
	
	var result strings.Builder
	start := 0
	
	for i := 0; i < n; i++ {
		remainingNeeded := n - i - 1
		end := bankLen - remainingNeeded
		
		maxDigit := bank[start]
		maxPos := start
		for j := start + 1; j < end; j++ {
			if bank[j] > maxDigit {
				maxDigit = bank[j]
				maxPos = j
			}
		}
		
		result.WriteByte(maxDigit)
		start = maxPos + 1
	}
	
	val, _ := strconv.ParseInt(result.String(), 10, 64)
	return val
}

func solve(inputData string) (string, string) {
	lines := strings.Split(strings.TrimSpace(inputData), "\n")
	
	var part1Sum int64 = 0
	var part2Sum int64 = 0
	
	for _, line := range lines {
		bank := strings.TrimSpace(line)
		if bank == "" {
			continue
		}
		
		part1Sum += findLargestSubsequence(bank, 2)
		
		if len(bank) >= 12 {
			part2Sum += findLargestSubsequence(bank, 12)
		}
	}
	
	return fmt.Sprintf("%d", part1Sum), fmt.Sprintf("%d", part2Sum)
}

func main() {
	data, err := ioutil.ReadFile("../data/input.txt")
	if err != nil {
		panic(err)
	}
	part1, part2 := solve(string(data))
	fmt.Printf("Part 1: %s\n", part1)
	fmt.Printf("Part 2: %s\n", part2)
}

File: days/day_03/solutions/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;

sub find_largest_subsequence {
    my ($bank, $n) = @_;
    return 0 if length($bank) < $n;
    
    my @result = ();
    my $start = 0;
    my $bank_len = length($bank);
    
    for my $i (0..$n-1) {
        # How many digits we still need after this one
        my $remaining_needed = $n - $i - 1;
        # The last index we can choose from
        my $end = $bank_len - $remaining_needed;
        
        # Find the maximum digit in the range [start, end)
        my $max_digit = substr($bank, $start, 1);
        my $max_pos = $start;
        for my $j ($start+1..$end-1) {
            my $digit = substr($bank, $j, 1);
            if ($digit gt $max_digit) {
                $max_digit = $digit;
                $max_pos = $j;
            }
        }
        
        push @result, $max_digit;
        $start = $max_pos + 1;
    }
    
    return int(join('', @result));
}

sub solve {
    my ($input_data) = @_;
    my @lines = split /\n/, $input_data;
    
    my $part1_sum = 0;
    my $part2_sum = 0;
    
    for my $line (@lines) {
        $line =~ s/^\s+|\s+$//g;
        next unless $line;
        my $bank = $line;
        
        # Part 1: largest 2-digit number
        $part1_sum += find_largest_subsequence($bank, 2);
        
        # Part 2: largest 12-digit number
        if (length($bank) >= 12) {
            $part2_sum += find_largest_subsequence($bank, 12);
        }
    }
    
    return ($part1_sum, $part2_sum);
}

open my $fh, '<', "../data/input.txt" or die "Cannot open file: $!";
my $data = do { local $/; <$fh> };
close $fh;

my ($part1, $part2) = solve($data);
print "Part 1: $part1\n";
print "Part 2: $part2\n";
